docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1056 @@
+
+```cs
+public unsafe class Benchmarks
+{
+    private void* _pointer;
+
+    [Params(6, 32, 1024)] // test various sizes
+    public uint Size;
+
+    [GlobalSetup]
+    public void Setup()
+    {
+        _pointer = NativeMemory.AlignedAlloc(byteCount: Size * sizeof(int), alignment: 32);
+        NativeMemory.Clear(_pointer, byteCount: Size * sizeof(int)); // ensure it's all zeros, so 1 is never found
+    }
+
+    [Benchmark]
+    public bool Contains()
+    {
+        ReadOnlySpan<int> buffer = new (_pointer, (int)Size);
+        return buffer.Contains(1);
+    }
+
+    [GlobalCleanup]
+    public void Cleanup() => NativeMemory.AlignedFree(_pointer);
+}
+```
+
+Sample results (please mind the AVX2, AVX and SSE4.2 information printed in the summary):
+
+```ini
+BenchmarkDotNet=v0.13.5, OS=Windows 11 (10.0.22621.1413/22H2/2022Update/SunValley2)
+AMD Ryzen Threadripper PRO 3945WX 12-Cores, 1 CPU, 24 logical and 12 physical cores
+.NET SDK=8.0.100-alpha.1.22558.1
+  [Host]    : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2
+  Scalar    : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT
+  Vector128 : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX
+  Vector256 : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2
+```
+
+```
+|   Method |       Job | Size |       Mean |    StdDev | Ratio | Code Size |
+|--------- |---------- |----- |-----------:|----------:|------:|----------:|
+| Contains |    Scalar | 1024 | 143.844 ns | 0.6234 ns |  1.00 |     206 B |
+| Contains | Vector128 | 1024 | 104.544 ns | 1.2792 ns |  0.73 |     335 B |
+| Contains | Vector256 | 1024 |  55.769 ns | 0.6720 ns |  0.39 |     391 B |
+```
+
+The results should be very stable (flat distributions), but on the other hand we are measuring the performance of the best case scenario (the input is large and its entire contents are searched through, as the value is never found).
+
+Explaining benchmark design guidelines is outside of the scope of this document, but we have a [dedicated document](https://github.com/dotnet/performance/blob/main/docs/microbenchmark-design-guidelines.md#benchmarks-are-not-unit-tests) about it. To make a long story short, **you should benchmark all scenarios that are realistic for your production environment**, so your customers can actually benefit from your improvements.
+
+##### Memory randomization
+
+The alternative is to enable memory randomization. Before every iteration, the harness is going to allocate random-size objects, keep them alive and re-run the setup that should allocate the actual memory.
+
+You can read more about it [here](https://github.com/dotnet/BenchmarkDotNet/pull/1587). It requires an understanding of what distribution is and how to read it. It's also out of scope of this document, but [Pro .NET Benchmarking](https://aakinshin.net/prodotnetbenchmarking/) has two chapters dedicated to statistics and can help you get a very good understanding of the subject.
```

Comment:
I'm not sure what our guidelines are for plugging for-pay content.

Reply:
perhaps "a book on statistics, such as .." ? Maybe there's another one to mention as an alternative.